### ISSUE TITLE
Training loss increases instead of deacreasing

### ISSUE THREAD

**Issue by utkuumur** [pytorch/xla#2482]
Hello,

I am experiencing an odd problem. I can train my model on GPU without a problem, training loss decreases slowly. But when I train my model on TPU, training loss increases quickly instead of decreasing slowly. 

I am training a Bert model using triplet networks. Model returns the vector corresponding to the "CLS " token of the given sentences, so after I retrieve three vectors, I use triplet loss with cosine similarity as the distance metric.

I divided my dataset into 20 chunks and load one of them at each epoch because of memory issues, but loss decreases with each batch. I could not understand why model performance decreases when training on TPU contrary to the training on GPU. I used TPU software version  "pytorch-1.5". I shared the training script below. Does anyone have any clue why this happens? Thanks in advance.

    import torch
    import torch.nn.functional as F
    import torch_xla
    import torch_xla.core.xla_model as xm
    import torch_xla.debug.metrics as met
    import torch_xla.distributed.xla_multiprocessing as xmp
    import torch_xla.distributed.data_parallel as dp
    import torch_xla.distributed.parallel_loader as pl
    from tqdm.autonotebook import tqdm, trange
    from torch.utils.data import DataLoader, RandomSampler, SequentialSampler, TensorDataset

    import transformers
    from sentence_transformers import SentenceTransformer, SentencesDataset, models, losses
    from sentence_transformers.readers import InputExample, TripletReader
    from sentence_transformers.util import batch_to_device

    import os
    import logging
    import math

    logger = logging.getLogger(__name__)

    class TripletLoss(torch.nn.Module):

        def __init__(self, margin=1.0, distance='l2'):
            super(TripletLoss, self).__init__()
            self.margin = margin
            self.distance = distance

        def forward(self, anchor, pos, neg):
            if self.distance == "cosine":
                dis_pos = F.cosine_similarity(anchor, pos)
                dis_neg = F.cosine_similarity(anchor, neg)
                losses = F.relu(dis_neg - dis_pos + self.margin)
            elif self.distance == "l2":
                dis_pos = F.pairwise_distance(anchor, pos)
                dis_neg = F.pairwise_distance(anchor, neg)
                losses = F.relu(dis_pos - dis_neg + self.margin)

            return losses.mean()
     

    def load_and_cache_example(index, model, max_examples=0, n_chunk=100):
        data_dir = "PATH"
        cached_file = os.path.join(data_dir, "cached_bert_train_part{}.data".format(index))

        xm.master_print('loading part:{}'.format(index))
         
        if xm.is_master_ordinal():
            if not os.path.exists(cached_file):
                patent_reader = TripletReader(data_dir, s1_col_idx=3, s2_col_idx=4, s3_col_idx=5, has_header=True)
                print("master started tokenizing")
                train_dataset = SentencesDataset(reader.get_examples('train{}.tsv'.format(index), max_examples=max_examples), model, max_processes=1, n_chunk=100)
                torch.save(train_dataset, cached_file)
            else:
                train_dataset = torch.load(cached_file)
                print('data size len:', len(train_dataset))

        xm.rendezvous("tokenize")
        
        if not xm.is_master_ordinal():
            train_dataset = torch.load(cached_file)

        return train_dataset

    def get_data_loader(index, device, model, batch_size = 8):
        dataset = load_and_cache_example(index,  model) 
        train_sampler = None 
        
        if xm.xrt_world_size() > 1:
            train_sampler = torch.utils.data.distributed.DistributedSampler(
                dataset,
                num_replicas=xm.xrt_world_size(),
                rank=xm.get_ordinal(),
                shuffle=True)
        
        train_dataloader = DataLoader(dataset, batch_size=batch_size, sampler=train_sampler, shuffle=False, collate_fn=model.smart_batching_collate)
        para_loader = pl.ParallelLoader(train_dataloader, [device])

        return para_loader 
        

    def _mp_fn(index):
        torch.set_default_tensor_type('torch.FloatTensor')

        logging.basicConfig(
            format="%(asctime)s - %(levelname)s - %(name)s -   %(message)s",
            datefmt="%m/%d/%Y %H:%M:%S",
            level=logging.INFO 
            )

        device = xm.xla_device()
        word_embedding_model = models.BERT('allenai/scibert_scivocab_uncased', max_seq_length=510, do_lower_case=True)

        pooling_model = models.Pooling(word_embedding_model.get_word_embedding_dimension(),
            pooling_mode_mean_tokens=False,
            pooling_mode_cls_token=True,
            pooling_mode_max_tokens=False)

        model = SentenceTransformer(modules=[word_embedding_model, pooling_model])
        model._target_device = device
        model = model.to(device)
        loss_fn = TripletLoss(distance="cosine", margin=0.5)
        loss_fn = loss_fn.to(device)
     
        tr_loss = train(model, loss_fn, device)
        xm.master_print('finished training')
        xm.master_print(met.metrics_report())

    def train(model, loss_fn, device):
        dataset_size = 20
        batch_size = 8
        len_dataset = 100000

        optimizer_class = transformers.AdamW
        optimizer_params = {'lr': 5e-6, 'eps': 1e-6, 'correct_bias': False}
        max_grad_norm = 1

        num_train_steps = dataset_size * len_dataset // batch_size // xm.xrt_world_size()
        warmup_steps = 100#math.ceil(num_train_steps *  0.1)
        xm.master_print("warm up steps: {}".format(warmup_steps))
        
        param_optimizer = list(model.named_parameters())
        no_decay = ['bias', 'LayernNorm.bias', 'LayerNorm.weight']

        optimizer_grouped_parameters = [
            {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
            {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
        ]

        optimizer = optimizer_class(optimizer_grouped_parameters, **optimizer_params)
        scheduler = model._get_scheduler(optimizer, scheduler="WarmupLinear", warmup_steps=warmup_steps, t_total=num_train_steps)

        global_step = 0
        loss = 0.0

        len_loader = len_dataset // batch_size // xm.xrt_world_size()

        logging.info("starting training")
        for epoch in trange(dataset_size, desc="Dataset", disable= not xm.is_master_ordinal()):
            para_loader = get_data_loader(epoch, device, model, batch_size=batch_size)
            data_iterator = para_loader.per_device_loader(device)
            
            model.train()

            for step in trange(len_loader, desc="Iteration", disable=not xm.is_master_ordinal()):
                data = next(data_iterator)          
                features, labels = batch_to_device(data, device)            

                anchor = model(features[0])['sentence_embedding']
                positive = model(features[1])['sentence_embedding']
                negative = model(features[2])['sentence_embedding']

                loss_value = loss_fn(anchor, positive, negative)
                loss_value.backward()
                torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)

                global_step += 1
                loss += loss_value.item()

                if global_step % 500 ==  0:
                    logger.info("Loss at {} global_step is: {}".format(global_step, loss / global_step))
            
                xm.optimizer_step(optimizer)
                scheduler.step()
                optimizer.zero_grad()
                

            
            xm.rendezvous("finish part")
            xm.master_print('Datset part {}  is finished by all process'.format(epoch))

            if (epoch + 1)  % 5 == 0:
                print('saving model')
                save(model, 'PATH/part{}/'.format((epoch+1)))
                xm.rendezvous("save model")

        
        if xm.is_master_ordinal():
            print('training is finished saving the model')
            save(model, "PATH/2m/")

        xm.rendezvous("save model")
        
        return loss
        
        
    def save(model, output_dir):
        if xm.is_master_ordinal():
            model._first_module().bert.config.save_pretrained(output_dir)
            model.tokenizer.save_pretrained(output_dir)
        xm.save(model._first_module().bert.state_dict(), output_dir + "pytorch_model.bin")

    if __name__ == '__main__':
        xmp.spawn(_mp_fn, args=(), nprocs=8, start_method='fork')

**Comment by jysohn23** [comment]
What version of the huggingface/transformers library are you using? Can you make sure this commit is included? https://github.com/huggingface/transformers/pull/6488

Also, just to make sure can you use our latest release `pytorch-1.6`?

**Comment by utkuumur** [comment]
I installed the transformers library from source 10 days ago. I switched to `pytorch-1.6` as you suggested and installed the `transformers=3.0.1` instead of the latest version and the problem seems to be solved.  I am not sure the commit you referenced is included but I am not using trainer so I think it is not relevant. I first tried with the latest version of the transformers library but it got stuck at first batch (not second) for 10 minutes, I tried restarting but that did not solve the issue, that is why I switched to version 3.0.1.  I also changed ` para_loader = pl.ParallelLoader(train_dataloader, [device])` line to 
`para_loader =l.MpDeviceLoader(train_dataloader, device)`. This is probably not necessary. 

Strangely, training performance degraded a bit (from 12 min to 18min per 100k example) (without loss.item()). This is still almost 2x faster than 4x Titan RTX.  Thanks for your help. I think this issue can be closed.
